### PR TITLE
Fabo/2288 handle txs async

### LINKED ISSUE #2288: Handle transactions included timeout
**Describe the bug**

If I send a transaction and it isn't included in a block right away, we get a timeout error. But the transaction actually is already in the mempool. This is dangerous, as people might try to send the transaction again. Currently this will cause an error as the tx is in mempool. Still it is confusing.

SDK related issue: https://github.com/cosmos/cosmos-sdk/issues/3875

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: 
 - Browser: 
 - Version: 

**Smartphone (please complete the following information):**
 - Device: 
 - OS: 
 - Browser 
 - Version 

**Additional context**

### PR DESCRIPTION
**Pull request by faboweb**
Closes #2288 

**Description:**

Thank you! 🚀

---

For contributor:

- [ ] Added entries in `PENDING.md` with issue # and GitHub username
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

File: app/src/renderer/vuex/modules/send.js [hunk 1]
```diff
@@ -179,12 +194,37 @@ export default ({ node }) => {
 
       // broadcast transaction with signatures included
       const signedTx = state.externals.createSignedTx(tx, signature)
-      const body = state.externals.createBroadcastBody(signedTx)
-      const res = await node.postTx(body).catch(handleSDKError)
+      const body = state.externals.createBroadcastBody(signedTx, `sync`)
+
```
Comment (fedekunze) [inline app/src/renderer/vuex/modules/send.js]:
```suggestion
```

File: app/src/renderer/vuex/modules/send.js [hunk 2]
```diff
@@ -179,12 +194,37 @@ export default ({ node }) => {
...
+        await new Promise(resolve => setTimeout(resolve, state.txQueryTimeout))
+        try {
+          await state.node.tx(txHash)
+          success = true
+        } catch (err) {
+          // tx wasn't included in a block yet
+        }
+      }
+      if (iterations <= 0) {
+        throw new Error(`The transaction was still not included in a block. We can't say for certain it will be included in the future.`)
```
Comment (fedekunze) [inline app/src/renderer/vuex/modules/send.js]: shouldn't this be a transaction pending on the Transactions page instead ? that seems like a more clean implementation to me (see designs from Zplin).

File: PENDING.md [hunk 3]
```diff
@@ -0,0 +1 @@
+[Changed] Poll for tx to be included in a block if it isn't right away @faboweb
```
Comment (fedekunze) [inline PENDING.md]: issue number

File: app/src/renderer/vuex/modules/send.js (hunk 2 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/send.js]: Step by step by step ;)

File: app/src/renderer/scripts/wallet.js [hunk 4]
```diff
@@ -173,9 +173,10 @@ export function createSignedTx(tx, signature) {
 }
 
 // the broadcast body consists of the signed tx and a return type
-export function createBroadcastBody(signedTx) {
+// returnType can be block (inclusion in block), async (right away), sync (after checkTx has passed)
+export function createBroadcastBody(signedTx, returnType = `block`) {
   return JSON.stringify({
     tx: signedTx,
```
Comment (sabau) [inline app/src/renderer/scripts/wallet.js]: just little nicety: if you want to rename signedTx to tx we can simplify line `179`

File: app/src/renderer/vuex/modules/send.js [hunk 5]
```diff
@@ -179,12 +194,36 @@ export default ({ node }) => {
...
+    },
+    // wait for inclusion of a tx in a block
+    async queryTxInclusion({ state }, txHash) {
+      let success = false
+      let iterations = state.txQueryIterations // 30 * 2s = 60s max waiting time
+      while (!success && iterations-- > 0) {
+        await new Promise(resolve => setTimeout(resolve, state.txQueryTimeout))
+        try {
+          await state.node.tx(txHash)
+          success = true
```
Comment (sabau) [inline app/src/renderer/vuex/modules/send.js]: I would try to simplify a by removing the variable `success` and replacing it here with a `break` + removing the `!success` and line `213`

Comment (sabau) [reply app/src/renderer/vuex/modules/send.js]: and moving the await inside the catch, so the very first time it won't wait (one failure extra) but it looks as a trade-off to improve a bit the easiness to read this loop

File: test/unit/specs/store/send.spec.js [hunk 6]
```diff
@@ -542,6 +542,74 @@ describe(`Module: Send`, () => {
...
+
+        expect(dispatch).toHaveBeenCalledWith(`queryTxInclusion`, `acbdefghi`)
+      })
+
+      it(`should poll txs block inclusion`, async () => {
+        jest.useRealTimers()
+
+        const node = {
+          tx: jest.fn()
+            .mockImplementationOnce(() => Promise.reject(new Error(`404`)))
```
Comment (sabau) [inline test/unit/specs/store/send.spec.js]: love this test

File: app/src/renderer/scripts/wallet.js (hunk 4 above)
Comment (faboweb) [reply app/src/renderer/scripts/wallet.js]: I thought that naming the argument `signedTx` is a good help for the developer to know that he needs to pass the signed tx in there. What do you think?

File: app/src/renderer/vuex/modules/send.js (hunk 5 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/send.js]: nice ideas

File: app/src/renderer/scripts/wallet.js (hunk 4 above)
Comment (sabau) [reply app/src/renderer/scripts/wallet.js]: That's true, this name is more informative!

**Pull request merged**